### ISSUE TITLE
Navigation doesn't fire from within a function which returns a promise or an observable

### ISSUE THREAD

**Issue by kostyanet** [angular/angular#17933]
## I'm submitting a ...
<pre><code>
[ ] Regression (behavior that used to work and stopped working in a new release)
[ x] Bug report 
[ ] Feature request
[ ] Documentation issue or request
</code></pre>

## Current behavior
If `showPopup` method returns a Promise or an Observable, navigation doesn't work properly: it fires only after the timer resolves. (The timer here is a substitute for some async logic). If it returns a primitive or does not return at all - navigations fires well.

## Expected behavior
Expect navigation to fire immediately regardless of the type being returned.

## Minimal reproduction of the problem with instructions
An excerpt from my popup service:
```
showPopup() {
    this.router.navigate([{ // <--------------------- navigation doesn't fire as expected
        outlets: { popup: ['popup'] },
    }]);

    return new Promise((resolve) => {
        setTimeout(() => {
            resolve();
        }, 2000);
    });
}
```

## Please tell us about your environment
<pre><code>
Angular version: 4.2.2
Browser:
- [ x] Chrome (iOS) version 59.0.3071.115 64-bit

**Comment by jasonaden** [comment]
I'm sorry but reported issues require a plunker reproducing the problem.

If this issue persists, please create a plunker using this template and describe the difference between the expected and current behavior and create a new issue: http://plnkr.co/edit/tpl:AvJOMERrnz94ekVua0u5?p=catalogue
